jax/lax_linalg.py

```diff
@@ -245,37 +247,89 @@ def _eigh_cpu_gpu_translation_rule(syevd_impl, c, operand, lower):
+  #   eigensensitivity with repeated roots. AIAA Journal 32, 1500-1506 (1994).
+  # Or look up "degenerate perturbation theory" in physics.
+  #
+  # Note that this approach does *not* correctly handle high order degeneracies,
+  # i.e., the case of degenerate eigenvalue derivatives. This means you cannot
+  # expect to get the right answer if you do higher order differentiation.
+  # Fixing this would require native support for higher order forwards-mode
+  # differentiation in JAX (https://github.com/google/jax/pull/1185).
+
   a, = primals
   a_dot, = tangents
 
-  v, w = eigh_p.bind(symmetrize(a), lower=lower)
+  a_sym = symmetrize(a)
+  v, w = eigh_p.bind(
+    a_sym, lower=lower, handle_degeneracies=handle_degeneracies)
 
   # for complex numbers we need eigenvalues to be full dtype of v, a:
   w = w.astype(a.dtype)
-  eye_n = np.eye(a.shape[-1], dtype=a.dtype)
-  # carefully build reciprocal delta-eigenvalue matrix, avoiding NaNs.
-  Fmat = np.reciprocal(eye_n + w[..., np.newaxis, :] - w[..., np.newaxis]) - eye_n
-  # eigh impl doesn't support batch dims, but future-proof the grad.
+  epsilon = 10 * np.finfo(a.dtype).resolution
+
+  deltas = w[..., np.newaxis, :] - w[..., np.newaxis]
+  same_subspace = (abs(deltas) < epsilon
+                   if handle_degeneracies
+                   else np.eye(a.shape[-1], dtype=bool))
+  # Note: if not handling degeneracies, this intentionally will result in
+  # nan/inf if there happen to be degenerate eigenvalues.
+  Fmat = np.where(same_subspace, 0.0, 1.0 / deltas)
+
   dot = lax.dot if a.ndim == 2 else lax.batch_matmul
-  vdag_adot_v = dot(dot(_H(v), a_dot), v)
-  dv = dot(v, np.multiply(Fmat, vdag_adot_v))
-  dw = np.diagonal(vdag_adot_v, axis1=-2, axis2=-1)
+  vdag_adot = dot(_H(v), a_dot)
+  vdag_adot_v = dot(vdag_adot, v)
+  C = Fmat * vdag_adot_v
+  dv_non_degenerate = dot(v, C)
+
+  if handle_degeneracies:
+    # Diagonalize the perturbation in any degenerate subspaces.
+    A, dw = eigh_p.bind(vdag_adot_v * same_subspace, lower=lower,
+                        handle_degeneracies=handle_degeneracies)
+
+    # Reorder these into sorted order of the original eigenvalues.
+    # TODO(shoyer): consider rewriting with an explicit loop over degenerate
+    # subspaces instead?
+    v2 = dot(v, A)
+    w2 = np.einsum('...ij,...jk,...ki->...i', _H(v2), a_sym, v2).real
+    order = np.argsort(w2, axis=-1)
+    A = np.take_along_axis(A, order[..., np.newaxis, :], axis=-1)
+    dw = np.take_along_axis(dw, order, axis=-1)
+
+    deltas_dot = dw[..., np.newaxis, :] - dw[..., np.newaxis]
+    same_dot_subspace = abs(deltas_dot) < epsilon
+    # If there are still some degenerate eigenvalue derivatives, the choice of
+    # basis is arbitrary (up to first order perturbations), so it's safe to set
+    # these terms in C_dot to 0.
+    Fmat_dot = np.where(same_dot_subspace, 0.0, 1.0 / deltas_dot)
+    vdag_adot_dv_non_degen = dot(vdag_adot, dv_non_degenerate)
+    C_dot = Fmat_dot * vdag_adot_dv_non_degen
```

Comment:
This is wrong. I am combining terms calculated in different bases (the original eigenbasis of `a` vs whatever diagonalizes `a_dot` as well in degenerate subspaces).